File: sp/engine/wat_runner.py
#!/usr/bin/env python3
"""
Statute & Precedent — WAT Runner (Python wrapper)
==================================================
Calls the Node.js WAT runner, parses results, and stores
web audit findings in the compliance SQLite database.

Usage:
    python3 wat_runner.py                  # Audit all companies
    python3 wat_runner.py --company alieninc  # Audit one company
    python3 wat_runner.py --report         # Just show latest report
"""

import json
import subprocess
import sys
import sqlite3
from pathlib import Path
from datetime import datetime

ENGINE_DIR = Path(__file__).parent
DB_PATH = ENGINE_DIR / "compliance.db"
REPORTS_DIR = ENGINE_DIR / "wat_reports"
COMBINED_REPORT = REPORTS_DIR / "combined_report.json"

# Company ID to name mapping — must match wat_runner.js COMPANIES keys
COMPANY_NAMES = {
    "alieninc":         {"name": "Alien.Inc",                "url": "https://alieninc.tech/"},
    "1609holdings":     {"name": "1609 Holdings",           "url": "https://1609.alieninc.tech"},
    "exosphere":        {"name": "Exosphere",               "url": "https://exosphere.alieninc.tech"},
    "panteon":        {"name": "Panteon",               "url": "https://panteon.alieninc.tech"},
    "kmt":              {"name": "KMT Consulting Group",    "url": "https://kmt.alieninc.tech"},
    "thedailyartcult":  {"name": "The Daily Art Cult",      "url": "https://thedailyartcult.lol"},
    "alcantara":        {"name": "St. Alcantara Foundation","url": "https://stalcantara.alieninc.tech"},
}


def init_web_audits_table(conn):
    """Create the web_audits table if it doesn't exist."""
    c = conn.cursor()
    c.execute("""
        CREATE TABLE IF NOT EXISTS web_audits (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            scan_date TEXT NOT NULL,
            company_id TEXT NOT NULL,
            cookies_total INTEGER DEFAULT 0,
            cookies_first_party INTEGER DEFAULT 0,
            cookies_third_party INTEGER DEFAULT 0,
            cookies_session INTEGER DEFAULT 0,
            cookies_persistent INTEGER DEFAULT 0,
            trackers_total INTEGER DEFAULT 0,
            tracker_lists TEXT DEFAULT '[]',
            hosts_total INTEGER DEFAULT 0,
            hosts_first_party INTEGER DEFAULT 0,
            hosts_third_party INTEGER DEFAULT 0,
            https_enabled INTEGER DEFAULT 0,
            https_redirect INTEGER DEFAULT 0,
            security_headers_score INTEGER DEFAULT 0,
            security_headers_missing TEXT DEFAULT '[]',
            localstorage_items INTEGER DEFAULT 0,
            web_compliance_score INTEGER DEFAULT 0,
            raw_report TEXT DEFAULT '{}'
        )
    """)
    c.execute("""
        CREATE INDEX IF NOT EXISTS idx_web_audits_company
        ON web_audits (company_id, scan_date)
    """)
    conn.commit()


def compute_web_compliance_score(audit):
    """Compute a 0-100 web compliance score from audit data.
    
    NOTE: Point deductions are based on regulatory requirements where applicable.
    Where no specific regulation mandates a technical measure, it is labeled as
    "best practice" rather than a compliance requirement.
    """
    score = 100

    # HTTPS check — Art. 32 recommends encryption as appropriate technical measure
    if not audit.get('https', {}).get('https_support', False):
        score -= 10

    # Third-party cookies — ePrivacy Directive Art. 5(3) requires consent
    cookies = audit.get('cookies', [])
    third_party = [c for c in cookies if not c.get('firstPartyStorage', True)]
    if len(third_party) > 0:
        score -= min(15, len(third_party) * 5)

    # Trackers — GDPR Art. 6 requires lawful basis for tracking
    beacons = audit.get('beacons', [])
    if len(beacons) > 0:
        score -= min(20, len(beacons) * 5)

    # Security headers — Art. 32 / Art. 5(1)(f) integrity and confidentiality
    headers = audit.get('security_headers', {})
    critical_headers = ['content-security-policy', 'strict-transport-security',
                        'x-frame-options', 'x-content-type-options']
    missing = [h for h in critical_headers if not headers.get(h)]
    if missing:
        score -= min(15, len(missing) * 5)

    # localStorage — Art. 5(1)(c) data minimisation principle
    ls = audit.get('localstorage', [])
    if len(ls) > 10:
        score -= min(5, (len(ls) - 10))

    # Third-party hosts — Art. 13(1)(e) requires recipient disclosure
    hosts = audit.get('hosts', [])
    third_party_hosts = [h for h in hosts if not h.get('firstParty', True)]
    if len(third_party_hosts) > 5:
        score -= min(5, (len(third_party_hosts) - 5))

    return max(0, score)


def run_wat_scan(company_id=None):
    """Run the Node.js WAT runner and store results."""
    cmd = ['node', str(ENGINE_DIR / 'wat_runner.js')]
    if company_id:
        info = COMPANY_NAMES.get(company_id)
        if not info:
            print(f"Unknown company: {company_id}")
            return None
        cmd.extend(['--url', info['url']])

    print("Running EDPB WAT web audit...")
    result = subprocess.run(cmd, capture_output=True, text=True, timeout=300)
    if result.returncode != 0:
        print(f"WAT runner error: {result.stderr}")
        return None

    print(result.stdout)
    return load_and_store_results()


def load_and_store_results():
    """Load the combined report and store in the database."""
    if not COMBINED_REPORT.exists():
        print("No WAT report found. Run the scanner first.")
        return None

    with open(COMBINED_REPORT) as f:
        report = json.load(f)

    conn = sqlite3.connect(str(DB_PATH))
    init_web_audits_table(conn)
    c = conn.cursor()
    scan_date = report.get('report_date', datetime.now().isoformat())
    scores = {}

    for company_id, audit in report.get('companies', {}).items():
        if 'error' in audit:
            print(f"  {company_id}: Error — {audit['error']}")
            continue

        score = compute_web_compliance_score(audit)
        scores[company_id] = score

        cookies = audit.get('cookies', [])
        hosts = audit.get('hosts', [])
        beacons = audit.get('beacons', [])

        first_party_cookies = [c for c in cookies if c.get('firstPartyStorage', True)]
        third_party_cookies = [c for c in cookies if not c.get('firstPartyStorage', True)]
        session_cookies = [c for c in cookies if c.get('session', False)]
        persistent_cookies = [c for c in cookies if not c.get('session', False)]

        first_party_hosts = [h for h in hosts if h.get('firstParty', True)]
        third_party_hosts = [h for h in hosts if not h.get('firstParty', True)]

        tracker_lists = list(set(b.get('listName', '') for b in beacons))

        headers = audit.get('security_headers', {})
        critical_headers = ['content-security-policy', 'strict-transport-security',
                            'x-frame-options', 'x-content-type-options',
                            'referrer-policy', 'permissions-policy']
        headers_score = sum(1 for h in critical_headers if headers.get(h))
        headers_missing = [h for h in critical_headers if not headers.get(h)]

        c.execute("""
            INSERT INTO web_audits
            (scan_date, company_id, cookies_total, cookies_first_party,
             cookies_third_party, cookies_session, cookies_persistent,
             trackers_total, tracker_lists,
             hosts_total, hosts_first_party, hosts_third_party,
             https_enabled, https_redirect,
             security_headers_score, security_headers_missing,
             localstorage_items, web_compliance_score, raw_report)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        """, (
            scan_date, company_id,
            len(cookies), len(first_party_cookies), len(third_party_cookies),
            len(session_cookies), len(persistent_cookies),
            len(beacons), json.dumps(tracker_lists),
            len(hosts), len(first_party_hosts), len(third_party_hosts),
            1 if audit.get('https', {}).get('https_support') else 0,
            1 if audit.get('https', {}).get('https_redirect') else 0,
            headers_score, json.dumps(headers_missing),
            len(audit.get('localstorage', [])),
            score, json.dumps(audit),
        ))

        cinfo = COMPANY_NAMES.get(company_id, {"name": company_id})
        if isinstance(cinfo, dict):
            display_name = cinfo.get('name', company_id)
        else:
            display_name = str(cinfo)
        print(f"  {display_name}: {score}% web compliance")

    conn.commit()
    conn.close()
    return scores


def get_latest_web_report():
    """Get the latest web audit report from the database."""
    if not DB_PATH.exists():
        return None

    conn = sqlite3.connect(str(DB_PATH))
    c = conn.cursor()

    # Check if table exists
    c.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='web_audits'")
    if not c.fetchone():
        conn.close()
        return None

    c.execute("""
        SELECT company_id, cookies_total, cookies_third_party,
               trackers_total, hosts_total, hosts_third_party,
               https_enabled, security_headers_score, security_headers_missing,
               localstorage_items, web_compliance_score, scan_date
        FROM web_audits
        WHERE id IN (SELECT MAX(id) FROM web_audits GROUP BY company_id)
        ORDER BY company_id
    """)

    results = []
    for row in c.fetchall():
        results.append({
            'company_id': row[0],
            'company_name': COMPANY_NAMES.get(row[0], row[0]),
            'cookies_total': row[1],
            'cookies_third_party': row[2],
            'trackers_total': row[3],
            'hosts_total': row[4],
            'hosts_third_party': row[5],
            'https_enabled': bool(row[6]),
            'security_headers_score': row[7],
            'security_headers_missing': json.loads(row[8]) if row[8] else [],
            'localstorage_items': row[9],
            'web_compliance_score': row[10],
            'scan_date': row[11],
        })

    conn.close()
    return results


if __name__ == "__main__":
    if '--report' in sys.argv:
        results = get_latest_web_report()
        if results:
            print(f"\n{'Company':30s} {'Score':>6s} {'Cookies':>8s} {'3P':>4s} {'Track':>6s} {'HTTPS':>6s} {'Headers':>8s}")
            print('─' * 80)
            for r in results:
                print(f"{r['company_name']:30s} {r['web_compliance_score']:>5}% {r['cookies_total']:>7d} {r['cookies_third_party']:>4d} {r['trackers_total']:>5d} {'yes' if r['https_enabled'] else 'NO':>6s} {r['security_headers_score']:>7d}/6")
        else:
            print("No web audit data available. Run a scan first.")
    else:
        company = None
        if '--company' in sys.argv:
            company = sys.argv[sys.argv.index('--company') + 1]
        run_wat_scan(company)

File: sp/engine/wat_runner.js
#!/usr/bin/env node
/**
 * Statute & Precedent — WAT Runner
 * =================================
 * Puppeteer-based website auditor using the EDPB WAT's own
 * @ghostery/adblocker libraries for tracker detection.
 *
 * Collects: cookies, localStorage, beacons/trackers, HTTPS status,
 * security headers, and third-party traffic for each company site.
 *
 * Usage: node wat_runner.js [--output-dir <path>] [--url <url>]
 * Default output: ./wat_reports/
 */

const fs = require('fs');
const path = require('path');
const puppeteer = require('puppeteer-core');
const { FiltersEngine } = require('@ghostery/adblocker');

// ── Configuration ─────────────────────────────────────────────────────────────
const OUTPUT_DIR = process.argv.includes('--output-dir')
    ? process.argv[process.argv.indexOf('--output-dir') + 1]
    : path.join(__dirname, 'wat_reports');

const SINGLE_URL = process.argv.includes('--url')
    ? process.argv[process.argv.indexOf('--url') + 1]
    : null;

const CHROMIUM_PATH = process.env.CHROMIUM_PATH || process.argv.includes('--chromium')
    ? (process.argv.includes('--chromium') ? process.argv[process.argv.indexOf('--chromium') + 1] : process.env.CHROMIUM_PATH)
    : '/usr/bin/chromium';
const PAGE_TIMEOUT = 30000;
const WAIT_AFTER_LOAD = 5000;

const COMPANIES = {
    alieninc:         { name: 'Alien.Inc',                    url: 'https://alieninc.tech/' },
    'rousseau':   { name: 'Rousseau',               url: 'https://rousseau.alieninc.tech' },
    panteon:        { name: 'Panteon',                    url: 'https://panteon.alieninc.tech' },
    immanuel:       { name: 'Immanuel',                   url: 'https://immanuel.alieninc.tech' },
    centra:         { name: 'Centra',                     url: 'https://centra.alieninc.tech' },
    kmt:              { name: 'KMT Consulting Group',         url: 'https://kmt.alieninc.tech' },
    thedailyartcult:  { name: 'The Daily Art Cult',           url: 'https://thedailyartcult.lol' },
    alcantara:        { name: 'Alcantara Art Foundation',     url: 'https://alcantaraartfoundation.alieninc.tech' },
};

// ── Load EDPB WAT adblocker filter lists ──────────────────────────────────────
const WAT_ASSETS = path.join(__dirname, 'assets');
const blockerOptions = {
    debug: true,
    enableOptimizations: false,
    loadCosmeticFilters: false,
};

let blockers = {};
try {
    blockers = {
        'easyprivacy': FiltersEngine.parse(
            fs.readFileSync(path.join(WAT_ASSETS, 'easyprivacy.txt'), 'utf8'),
            blockerOptions
        ),
        'fanboy-annoyance': FiltersEngine.parse(
            fs.readFileSync(path.join(WAT_ASSETS, 'fanboy-annoyance.txt'), 'utf8'),
            blockerOptions
        ),
    };
    console.log(`  Loaded ${Object.keys(blockers).length} filter lists from EDPB WAT assets`);
} catch (e) {
    console.error(`  Warning: Could not load filter lists: ${e.message}`);
    console.error(`  Tracker detection will be limited.`);
}

// ── Audit a single URL ────────────────────────────────────────────────────────
async function auditUrl(browser, url, companyName) {
    const page = await browser.newPage();
    await page.setExtraHTTPHeaders({
        'X-AlienInc-Audit': 'statute',
    });
    const results = {
        url,
        company: companyName,
        audit_time: new Date().toISOString(),
        cookies: [],
        localstorage: [],
        beacons: [],
        hosts: [],
        https: {},
        security_headers: {},
        errors: [],
    };

    // Collect network requests for beacons + hosts
    const requests = [];
    const hostsSet = new Set();

    // Cookie collection from HTTP headers
    const httpCookies = [];

    page.on('request', (req) => {
        const parsed = new URL(req.url());
        if (parsed.protocol !== 'data:') {
            hostsSet.add(parsed.hostname);
        }
        requests.push({
            url: req.url(),
            method: req.method(),
            resourceType: req.resourceType(),
            headers: req.headers(),
            sourceUrl: req.frame()?.url() || url,
        });
    });

    page.on('response', async (res) => {
        const resUrl = res.url();
        // Remove trailing slash for matching
        const cleanUrl = resUrl.replace(/\/$/, '');
        const pageUrl = url.replace(/\/$/, '');

        // Collect Set-Cookie headers
        const headers = res.headers();
        const setCookie = headers['set-cookie'];
        if (setCookie) {
            const cookieStrings = Array.isArray(setCookie) ? setCookie : [setCookie];
            for (const raw of cookieStrings) {
                httpCookies.push({
                    raw,
                    source: 'Set-Cookie header',
                    url: resUrl,
                    type: 'Cookie.HTTP',
                });
            }
        }

        // Collect security headers from the main page response
        if (cleanUrl === pageUrl || resUrl === url || cleanUrl.endsWith(pageUrl) || pageUrl.endsWith(cleanUrl)) {
            results.security_headers = {
                'strict-transport-security': headers['strict-transport-security'] || null,
                'content-security-policy': headers['content-security-policy'] || null,
                'x-frame-options': headers['x-frame-options'] || null,
                'x-content-type-options': headers['x-content-type-options'] || null,
                'x-xss-protection': headers['x-xss-protection'] || null,
                'referrer-policy': headers['referrer-policy'] || null,
                'permissions-policy': headers['permissions-policy'] || null,
                'server': headers['server'] || null,
            };
        }
    });

    // Intercept document.cookie writes
    await page.evaluateOnNewDocument(() => {
        window.__wat_js_cookies = [];
        const origDescriptor = Object.getOwnPropertyDescriptor(Document.prototype, 'cookie');
        Object.defineProperty(document, 'cookie', {
            get: function() {
                return origDescriptor.get.call(this);
            },
            set: function(val) {
                window.__wat_js_cookies.push({ raw: val, type: 'Cookie.JS' });
                return origDescriptor.set.call(this, val);
            },
        });

        // Hook localStorage.setItem
        window.__wat_localstorage = [];
        const origSetItem = Storage.prototype.setItem;
        Storage.prototype.setItem = function(key, value) {
            try {
                window.__wat_localstorage.push({
                    host: window.location.hostname,
                    key,
                    value: String(value).substring(0, 500),
                });
            } catch(e) {}
            return origSetItem.call(this, key, value);
        };
    });

    try {
        // Navigate
        const response = await page.goto(url, {
            waitUntil: 'networkidle2',
            timeout: PAGE_TIMEOUT,
        });

        // Capture headers from the main response directly
        if (response) {
            const respHeaders = response.headers();
            results.security_headers = {
                'strict-transport-security': respHeaders['strict-transport-security'] || null,
                'content-security-policy': respHeaders['content-security-policy'] || null,
                'x-frame-options': respHeaders['x-frame-options'] || null,
                'x-content-type-options': respHeaders['x-content-type-options'] || null,
                'x-xss-protection': respHeaders['x-xss-protection'] || null,
                'referrer-policy': respHeaders['referrer-policy'] || null,
                'permissions-policy': respHeaders['permissions-policy'] || null,
                'server': respHeaders['server'] || null,
            };
        }

        // HTTPS check
        const finalUrl = page.url();
        results.https = {
            original_url: url,
            final_url: finalUrl,
            https_redirect: url.startsWith('http://') && finalUrl.startsWith('https://'),
            https_support: finalUrl.startsWith('https://'),
            status_code: response?.status() || 0,
        };

        // Wait for JS-set cookies and localStorage
        await new Promise(r => setTimeout(r, WAIT_AFTER_LOAD));

        // Collect JS cookies
        const jsCookies = await page.evaluate(() => window.__wat_js_cookies || []);
        const jsLocalstorage = await page.evaluate(() => window.__wat_localstorage || []);

        // Browser cookies
        const browserCookies = await page.cookies();

        // Parse all cookies
        results.cookies = parseCookies(browserCookies, httpCookies, jsCookies, finalUrl);
        results.localstorage = jsLocalstorage;

        // Match network requests against adblocker filter lists
        results.beacons = matchTrackers(requests);

        // Third-party hosts
        const pageHost = new URL(finalUrl).hostname;
        results.hosts = Array.from(hostsSet).map(h => ({
            hostname: h,
            firstParty: h === pageHost || h.endsWith('.' + pageHost),
        }));

        // Page metadata
        results.page_title = await page.title();

    } catch (e) {
        results.errors.push({ message: e.message, stack: e.stack });
    }

    await page.close();
    return results;
}

// ── Parse cookies from all sources ────────────────────────────────────────────
function parseCookies(browserCookies, httpCookies, jsCookies, pageUrl) {
    const seen = new Map();

    // Browser cookies (from Puppeteer cookie jar)
    for (const c of browserCookies) {
        const key = `${c.name}|${c.domain}|${c.path}`;
        seen.set(key, {
            name: c.name,
            value: c.value,
            domain: c.domain.replace(/^\./, ''),
            path: c.path,
            secure: c.secure,
            httpOnly: c.httpOnly,
            sameSite: c.sameSite || 'unspecified',
            session: !c.expires || c.expires === -1,
            expires: c.expires || -1,
            firstPartyStorage: isFirstParty(c.domain, pageUrl),
            source: 'browser',
        });
    }

    // HTTP Set-Cookie parsed
    for (const hc of httpCookies) {
        const parts = hc.raw.split(';')[0].split('=');
        if (parts.length >= 2) {
            const name = parts[0].trim();
            const key = `${name}|${new URL(hc.url).hostname}|/`;
            if (!seen.has(key)) {
                seen.set(key, {
                    name,
                    value: parts.slice(1).join('=').trim(),
                    domain: new URL(hc.url).hostname,
                    path: '/',
                    secure: hc.raw.toLowerCase().includes('secure'),
                    httpOnly: hc.raw.toLowerCase().includes('httponly'),
                    sameSite: extractSameSite(hc.raw),
                    session: hc.raw.toLowerCase().includes('session') || !hc.raw.toLowerCase().includes('expires'),
                    expires: -1,
                    firstPartyStorage: true,
                    source: 'http-header',
                });
            }
        }
    }

    // JS-set cookies
    for (const jc of jsCookies) {
        const parts = jc.raw.split(';')[0].split('=');
        if (parts.length >= 2) {
            const name = parts[0].trim();
            const key = `${name}||/`;
            if (!seen.has(key)) {
                seen.set(key, {
                    name,
                    value: parts.slice(1).join('=').trim(),
                    domain: new URL(pageUrl).hostname,
                    path: '/',
                    secure: pageUrl.startsWith('https'),
                    httpOnly: false,
                    sameSite: 'unspecified',
                    session: jc.raw.toLowerCase().includes('session') || !jc.raw.toLowerCase().includes('expires'),
                    expires: -1,
                    firstPartyStorage: true,
                    source: 'javascript',
                });
            }
        }
    }

    return Array.from(seen.values());
}

function isFirstParty(cookieDomain, pageUrl) {
    try {
        const pageHost = new URL(pageUrl).hostname;
        const domain = cookieDomain.replace(/^\./, '');
        return pageHost === domain || pageHost.endsWith('.' + domain) || domain.endsWith('.' + pageHost);
    } catch { return false; }
}

function extractSameSite(raw) {
    const lower = raw.toLowerCase();
    if (lower.includes('samesite=strict')) return 'Strict';
    if (lower.includes('samesite=lax')) return 'Lax';
    if (lower.includes('samesite=none')) return 'None';
    return 'unspecified';
}

// ── Match network requests against adblocker filter lists ─────────────────────
function matchTrackers(requests) {
    const trackerMap = new Map();

    for (const req of requests) {
        if (req.resourceType === 'document' || req.resourceType === '') continue;

        const sourceUrl = req.sourceUrl || '';
        const adblockerReq = {
            url: req.url,
            sourceUrl,
            type: req.resourceType,
        };

        for (const [listName, blocker] of Object.entries(blockers)) {
            try {
                const { match, filter } = blocker.match(adblockerReq);
                if (match) {
                    const parsed = new URL(req.url);
                    const key = `${parsed.hostname}${parsed.pathname}`;
                    if (!trackerMap.has(key)) {
                        trackerMap.set(key, {
                            url: req.url,
                            hostname: parsed.hostname,
                            pathname: parsed.pathname,
                            filter: filter ? filter.toString() : '',
                            listName,
                            occurrences: 0,
                            resources: [],
                        });
                    }
                    const entry = trackerMap.get(key);
                    entry.occurrences++;
                    if (entry.resources.length < 5) {
                        entry.resources.push(req.url);
                    }
                }
            } catch (e) {
                // Skip malformed URLs
            }
        }
    }

    return Array.from(trackerMap.values()).sort((a, b) => b.occurrences - a.occurrences);
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
    console.log('\n  Statute & Precedent — WAT Runner');
    console.log('  ─────────────────────────────────────────');
    console.log('  EDPB Website Auditing Tool integration');
    console.log('  Using @ghostery/adblocker for tracker detection\n');

    // Ensure output directory
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });

    // Launch browser
    console.log('  Launching Chromium...');
    const browser = await puppeteer.launch({
        executablePath: CHROMIUM_PATH,
        headless: 'new',
        args: [
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-gpu',
            '--disable-dev-shm-usage',
        ],
    });
    console.log('  Browser ready.\n');

    const targets = SINGLE_URL
        ? { custom: { name: 'Custom URL', url: SINGLE_URL } }
        : COMPANIES;

    const allResults = {};

    for (const [id, info] of Object.entries(targets)) {
        console.log(`  Auditing ${info.name}...`);
        try {
            const result = await auditUrl(browser, info.url, info.name);
            allResults[id] = result;

            const cookieCount = result.cookies.length;
            const trackerCount = result.beacons.length;
            const hostCount = result.hosts.length;
            const httpsOk = result.https.https_support ? 'HTTPS' : 'HTTP';

            console.log(`    ${httpsOk} | ${cookieCount} cookies | ${trackerCount} trackers | ${hostCount} hosts`);

            // Save individual report
            const reportPath = path.join(OUTPUT_DIR, `${id}.json`);
            fs.writeFileSync(reportPath, JSON.stringify(result, null, 2));
        } catch (e) {
            console.error(`    Error: ${e.message}`);
            allResults[id] = { error: e.message };
        }
    }

    // Save combined report
    const combinedPath = path.join(OUTPUT_DIR, 'combined_report.json');
    const combined = {
        report_date: new Date().toISOString(),
        tool: 'EDPB WAT Runner (Puppeteer)',
        version: '2.0.1',
        companies: allResults,
    };
    fs.writeFileSync(combinedPath, JSON.stringify(combined, null, 2));

    await browser.close();

    console.log(`\n  Reports saved to ${OUTPUT_DIR}/`);
    console.log(`  Combined report: ${combinedPath}\n`);

    return combined;
}

main().catch(e => {
    console.error('Fatal error:', e);
    process.exit(1);
});
